### ISSUE TITLE
Event dispatching in Xish should trap errors

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2616]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)| @ralphm reported|
|-|-|
|Trac ID|trac#2616|
|Type|defect|
|Created|2007-05-01 12:25:56Z|

The dispatch of an event should not cause the exceptions from the callbacks to bubble up to the caller of `t.w.x.utility.EventDispatcher.dispatch`. Instead, such errors should probably be trapped, logged and discarded.

<details><summary>Searchable metadata</summary>

```
trac-id__2616 2616
type__defect defect
reporter__ralphm ralphm
priority__highest highest
milestone__Words_0_6 Words-0.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1178022356000000 1178022356000000
changetime__1178275659000000 1178275659000000
version__None None
owner__ 
cc__ralphm cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_1"></a>|@ralphm removed owner|
|-|-|

Implemented in [branches/xish-dispatch-2616.](branches/xish-dispatch-2616.)

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @ralphm|
|-|-|

For the sake of nitpicking, I'd prefer ```raise TestError()``` instead of ```raise TestError```. Otherwise that looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_3"></a>|@ralphm set status to `closed`|
|-|-|

(In [20182]) Trap and log exceptions raised in Xish observers.

Author: ralphm
Reviewer: therve
Fixes [#2616](https://github.com/twisted/twisted/issues/2616).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_4"></a>|@ralphm commented|
|-|-|

(In [20183]) Revert r20181. Test failing on some platforms.

Re [#2616](https://github.com/twisted/twisted/issues/2616).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

The try/except should be done in CallbackList.callback, not in dispatch. The test passed sometimes because the order of a dict is dependent of the platform, and in our tests cb.call was always run the first.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm commented|
|-|-|

Modified test to consistently fail on the previous implementation, moved trapping of exception within the loop that iterates the callbacks in a `CallbackList`.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set owner to @therve<br>@ralphm set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @ralphm|
|-|-|

Cool, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_10"></a>|@ralphm set status to `closed`|
|-|-|

(In [20185]) Trap and log exceptions raised in Xish observers.

Author: ralphm
Reviewer: therve
Fixes [#2616](https://github.com/twisted/twisted/issues/2616).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
